Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1

```diff
@@ -473,6 +479,92 @@ function Get-MsiAccessToken {
     }
 }
 
+function Get-SpnAccessTokenUsingCertificate {
+    param(
+        [Parameter(Mandatory=$true)] $endpoint
+    )
+
+    if ($script:certificateAccessToken -and $script:certificateAccessToken.expires_on) {
+        # there exists a token cache
+        $currentTime = [System.DateTime]::UtcNow
+        $tokenExpiryTime = $script:certificateAccessToken.expires_on.UtcDateTime
+        $timeDifference = $tokenExpiryTime - $currentTime
+
+        if ($timeDifference.Minutes -gt 5) {
+            Write-Verbose "Returning access token from cache."
+            return $script:certificateAccessToken
+        }
+    }
+
+    Write-Verbose "Fetching access token using client certificate."
+    
+    # load the ADAL library 
+    Add-Type -Path $PSScriptRoot\Microsoft.IdentityModel.Clients.ActiveDirectory.dll
+
+    $pemFileContent = $endpoint.Auth.Parameters.ServicePrincipalCertificate
+    $pfxFilePath, $pfxFilePassword = ConvertTo-Pfx -pemFileContent $pemFileContent
+    
+    $clientCertificate = Get-PfxCertificate -pfxFilePath $pfxFilePath -pfxFilePassword $pfxFilePassword
+
+    $servicePrincipalId = $endpoint.Auth.Parameters.ServicePrincipalId
+    $tenantId = $endpoint.Auth.Parameters.TenantId
+    $envAuthUrl = $script:defaultEnvironmentAuthUri
+    if($endpoint.Data.environmentAuthorityUrl) {
+        $envAuthUrl = $endpoint.Data.environmentAuthorityUrl
+    }
+
+    $envAuthUrl = Get-EnvironmentAuthUrl -endpoint $endpoint
+    $azureActiveDirectoryResourceId = Get-AzureActiverDirectoryResourceId -endpoint $endpoint
+    $authorityUrl = $envAuthUrl
+
+    $isADFSEnabled = $false
+    if ($endpoint.Data.EnableAdfsAuthentication -eq "true") {
+        $isADFSEnabled = $true
+    }
+
+    if (-not $isADFSEnabled) {
+        $authorityUrl = "$envAuthUrl$tenantId"
+    }
+
+    try {
+        $clientAssertionCertificate = New-Object Microsoft.IdentityModel.Clients.ActiveDirectory.ClientAssertionCertificate -ArgumentList $servicePrincipalId, $clientCertificate
+
+        $validateAuthority = -not $isADFSEnabled
```

Comment:
Why this?

Reply:
This is needed because authentication context currently doesn't validate authority for ADFS.